### ISSUE TITLE
unevaluatedProperties doesn't appear to be implemented

### ISSUE THREAD

**Issue by ssilverman** [networknt/json-schema-validator#276]
According to my testing, it doesn't appear that `unevaluatedProperties` from 2019-09 is implemented. If I'm correct, are there plans to add support?

**Comment by stevehu** [comment]
@ssilverman You are correct that it is not implemented yet. We implemented most of the validators but there are still some pending. I am wondering if you could submit a PR for this validator. Thanks.

**Comment by ssilverman** [comment]
Ah, I hadn't realized the 2019-09 support wasn't complete yet, that's why I filed this issue. I can't promise I have time at the moment, but I'm going to have a look at the code. Any implementation guidance for newcomers? Specifically, `unevaluatedProperties` needs to be evaluated after a few other properties are evaluated and collect info from them.

**Comment by stevehu** [comment]
@ssilverman The library is a flexible plugin structure. Here are some of the points that I can think of. 

* Create a new validator
* Register the validator to the ValidatorTypeCode
* There are two ways to carry the context information across multiple validators. ValidationContext or ThreadLocal. There are examples in the library about how these two options work.  

Let me know when you encounter any specific questions. Thanks.

**Comment by ssilverman** [comment]
I started here: https://github.com/ssilverman/json-schema-validator/commit/35a8889d03e9c8810bf1ace058dfe4be95793c53

I think I'll need some guidance when adding "peering into" the in-place applicators (`allOf`, `anyOf`, `oneOf`, and `not`) because they'll need some recursion, terminating at schemas.

**Comment by stevehu** [comment]
You are on the right path. Here is the document talking about the version.
 
https://github.com/networknt/json-schema-validator/blob/master/doc/specversion.md

**Comment by ssilverman** [comment]
I added the in-place applicator collection in the `UnevaluatedProperties` constructor, but I'm not 100% clear on how to go about adding `$ref` and in-place applicator results to `validate`.

See: https://github.com/ssilverman/json-schema-validator/commits/add-unevaluated-props

**Comment by stevehu** [comment]
@ssilverman To be honest, I am far behind as I am still learning the spec. I am wondering if this is already published or not. The only document I can find is this [issue](https://github.com/json-schema-org/json-schema-spec/issues/556). Let me know if you have any official links for us to learn. 

I looked at your code and it looks pretty complicated. I am wondering if you could create a document in the /doc folder to write down your implementation ideas. In this way, it might be easier for other developers to understand what you are trying to archive and help you if needed. Thanks.

**Comment by ssilverman** [comment]
It's published in the 2019-09 draft.

As for the complexity, I'm just following the existing code style for the other validators. "unevaluatedProperties" is similar to "additionalProperties", but it also has to look at any sibling "additionalProperties", "oneOf", "allOf", "anyOf", "not", and "ref" contents. How could I make it simpler? I don't yet know your code base that well.

**Comment by stevehu** [comment]
@ssilverman Thanks for the link. It is not very clear what needs to be done by reading for that paragraph :) My understanding is that we need to iterate all the siblings. Do we need to drill down to the siblings? Or we just need to evaluate it at the same level? I don't think there is a simpler way to do so. @prashanthjos has created a collector to collect useful info during the validation and it could iterate all validators. I am wondering if it could help. https://github.com/networknt/json-schema-validator/pull/268

**Comment by gjvoosten** [comment]
Since the new keyword `unevaluatedItems` (see https://json-schema.org/draft/2019-09/release-notes.html#applicator-vocabulary) is also not implemented, maybe take these together (and change the title of this issue to _unevaluatedProperties and unevaluatedItems keywords not implemented_)?

**Comment by stevehu** [comment]
@gjvoosten Good idea. I have changed the title to cover both keywords. Thanks.

**Comment by Gaff** [comment]
@ssilverman Just looking at your branch. Maybe it would be better to add an interface, something like:

```java
public boolean wouldHandleProperty(JsonNode node, String propertyName);
```

Then the `UnevaluatedPropertiesValidator` could call the other validators with `wouldHandleProperty()` to figure out which properties have been handled?

**Comment by ssilverman** [comment]
@Gaff I ended up writing my own validator :)

**Comment by stevehu** [comment]
@ssilverman I am assuming you add some dummy validators that doing nothing to map to these keywords.

**Comment by ssilverman** [comment]
@stevehu do you mean in my validator?

**Comment by stevehu** [comment]
@ssilverman I am sorry. I thought this thread is talking about the warning messages for the un-implemented keywords. Please ignore my message above. Thanks.

**Comment by Johnlon** [comment]
Any PR?

**Comment by fdutton** [comment]
Resolved in #716 and #749
